### PR TITLE
 Build gather-sysinfo

### PR DESCRIPTION
**Pull request by MarSik**
The gather-sysinfo binary is used by must-gather to collect
structured hardware information data about an OCP node.

This data is then also used by the Performance Profile Creator
also present in this repository.

Must-gather also needs the lspci binary from the pciutils package.

### CODE REVIEW

File: cmd/gather-sysinfo/output
Comment (jlojosnegros) [inline cmd/gather-sysinfo/output:1]: I think this file has been committed by mistake

File: cmd/gather-sysinfo/gather-sysinfo_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package main
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
```
Comment (ffromani) [inline cmd/gather-sysinfo/gather-sysinfo_suite_test.go:8]: we should have moved away from ginkgo for these tests long ago

File: cmd/gather-sysinfo/testSnapshot.tgz
Comment (jlojosnegros) [inline cmd/gather-sysinfo/testSnapshot.tgz:1]: this file also

File: cmd/gather-sysinfo/gather-sysinfo_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"io/ioutil"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
```
Comment (ffromani) [inline cmd/gather-sysinfo/gather-sysinfo_test.go:7]: ditto

**Review by ffromani — COMMENTED** [review]
importing `gather-sysinfo` into the tree may and should be the chance for a good and long-needed cleanup, but I guess we don't have time/capacity for that, do we?

File: cmd/gather-sysinfo/testSnapshot.tgz
Comment (MarSik) [reply cmd/gather-sysinfo/testSnapshot.tgz:1]: Part of the tests afaik.

File: cmd/gather-sysinfo/output
Comment (MarSik) [reply cmd/gather-sysinfo/output:1]: It is part of the test.

File: cmd/gather-sysinfo/gather-sysinfo_suite_test.go (hunk 1 above)
Comment (MarSik) [reply cmd/gather-sysinfo/gather-sysinfo_suite_test.go]: updated

File: cmd/gather-sysinfo/gather-sysinfo_test.go (hunk 2 above)
Comment (MarSik) [reply cmd/gather-sysinfo/gather-sysinfo_test.go]: updated

File: cmd/gather-sysinfo/testSnapshot.tgz
Comment (MarSik) [reply cmd/gather-sysinfo/testSnapshot.tgz:1]: You were right, removed.

File: cmd/gather-sysinfo/output
Comment (MarSik) [reply cmd/gather-sysinfo/output:1]: Removed.

File: go.mod [hunk 3]
```diff
@@ -160,6 +162,7 @@ require (
 
 // Pinned to kubernetes-1.26.2
 replace (
+	github.com/google/cadvisor => github.com/google/cadvisor v0.43.0
```
Comment (ffromani) [inline go.mod:165]: we previously consumed 0.46.0, I'd fix the code to use the newer version (should any change be needed) rather than jumping back

**Comment by ffromani** [comment]
thanks for the updates! next time I recommend considering a verbatim import + fixing in the subsequent commits in order to preserve a good audit trail. In some cases this is unpractical but in this case it could have been.

File: go.mod (hunk 3 above)
Comment (MarSik) [reply go.mod]: I just did, this was the reason I had to fix debug tools first.

File: cmd/gather-sysinfo/gather-sysinfo.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/jaypipes/ghw/pkg/snapshot"
+	"github.com/openshift-kni/debug-tools/pkg/knit/cmd"
+	"github.com/spf13/cobra"
+)
```
Comment (ffromani) [inline cmd/gather-sysinfo/gather-sysinfo.go:15]: can we reorder imports?

File: cmd/gather-sysinfo/gather-sysinfo.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return makeSnapshot(cmd, knitOpts, opts, args)
+		},
+		Args: cobra.NoArgs,
+	}
+	snap.Flags().StringVar(&opts.rootDir, "root", "", "pseudofs root - use this if running inside a container")
+	snap.Flags().StringVar(&opts.output, "output", "", "path to clone system information into")
+	snap.Flags().BoolVar(&opts.dumpList, "dump", false, "just dump the glob list of expected content and exit")
+	// use this to debug container behaviour
+	snap.Flags().IntVar(&opts.sleepTime, "sleep", 0, "amount of seconds to sleep once done, before exit")
```
Comment (ffromani) [inline cmd/gather-sysinfo/gather-sysinfo.go:49]: this raised concerns in review and we can probably replace anyway with something better. I'd remove it.

File: cmd/gather-sysinfo/gather-sysinfo_test.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+import (
+	"io/ioutil"
+	"k8s.io/utils/strings/slices"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/openshift-kni/debug-tools/pkg/knit/cmd"
+	"github.com/spf13/cobra"
```
Comment (ffromani) [inline cmd/gather-sysinfo/gather-sysinfo_test.go:11]: can we reorder imports?

File: cmd/gather-sysinfo/gather-sysinfo_test.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+	err := makeSnapshot(cmd, knitOpts, opts, args)
+	if err == nil {
+		t.Errorf("Failure was expected when running without --output argument.")
+	}
+	t.Log(err)
+
+	opts.output = "testSnapshot.tgz"
+
+	// Delete old snapshot if it exists
+	_ = os.Remove(opts.output) // ignore error
```
Comment (ffromani) [inline cmd/gather-sysinfo/gather-sysinfo_test.go:102]: we leaked this in the past, should probably be moved into a `defer` block

Comment (MarSik) [reply cmd/gather-sysinfo/gather-sysinfo_test.go:102]: Isn't the leak intentional? To allow debugging of failed tests?

Comment (ffromani) [reply cmd/gather-sysinfo/gather-sysinfo_test.go:102]: right but if so (sloppy me) it should be controllable from outside (flags, env vars...)

File: Makefile [hunk 8]
```diff
@@ -242,6 +242,19 @@ performance-profile-creator-tests: build-performance-profile-creator
 	@echo "Running Performance Profile Creator Tests"
 	hack/run-test.sh -t "test/e2e/performanceprofile/functests-performance-profile-creator" -p "--v -r --fail-fast --flake-attempts=2" -m "Running Functional Tests" -r "--junit-report=/tmp/artifacts"
 
+# Gather sysinfo binary for use in must-gather
+.PHONY: build-gather-sysinfo
+build-gather-sysinfo:
+	@echo "Building gather-sysinfo"
+	LDFLAGS="-s -w -X ${PACKAGE}/cmd/gather-sysinfo/version.Version=${REV} "; \
+	$(GO) build  -v $(LDFLAGS) -o $(OUT_DIR)/gather-sysinfo ./cmd/gather-sysinfo
```
Comment (jmencak) [inline Makefile:250]: Nit: formatting double spaces.

File: Makefile [hunk 9]
```diff
@@ -242,6 +242,19 @@ performance-profile-creator-tests: build-performance-profile-creator
...
+
+.PHONY: gather-sysinfo-tests
+gather-sysinfo-tests: build-gather-sysinfo
+	@echo "Running gather-sysinfo Tests"
+	$(GO) test  -v ./cmd/gather-sysinfo
```
Comment (jmencak) [inline Makefile:255]: Nit: formatting double spaces.  I guess copied from `build-performance-profile-creator`

**Review by jmencak — COMMENTED** [review]
The changes overall seem fine, but it is not clear to me how the collected information gets into must-gather.  The commit message only states:

    The gather-sysinfo binary is used by must-gather to collect
    structured hardware information data about an OCP node.

Can you expand on that a bit how must-gather gets that information?  Perhaps even in the commit message?  Saw some nits so I pointed them out.

File: Makefile (hunk 9 above)
Comment (MarSik) [reply Makefile:255]: Yeah, I assumed that was intentional :)

**Pull request merged**